README.md
# Quick JSON reponses in Laravel 

[![Latest Version on Packagist](https://img.shields.io/packagist/v/Witify/laravel-json-response.svg?style=flat-square)](https://packagist.org/packages/witify/laravel-json-response)
[![Build Status](https://img.shields.io/travis/Witify/laravel-json-response/master.svg?style=flat-square)](https://travis-ci.org/witify/laravel-json-response)
[![Maintainability](https://api.codeclimate.com/v1/badges/d536d6a0acff44433c69/maintainability)](https://codeclimate.com/github/Witify/laravel-json-response/maintainability)
[![Test Coverage](https://api.codeclimate.com/v1/badges/d536d6a0acff44433c69/test_coverage)](https://codeclimate.com/github/Witify/laravel-json-response/test_coverage)
[![Total Downloads](https://img.shields.io/packagist/dt/Witify/laravel-json-response.svg?style=flat-square)](https://packagist.org/packages/witify/laravel-json-response)

## Installation

You can install the package via composer:

```bash
composer require witify/laravel-json-response
```

## Usage

Add it to your controller:
``` php

use Witify\LaravelJsonReponse\JsonTrait;

class ApiController extends Controller {
    use JsonTrait;
}
```

## Available methods

| Method | Description |
| --- | --- |
| setStatusCode(int $statusCode) | Set the status code of the response |
| getStatusCode(int $statusCode) | Get the status code of the response |
| response($data, $headers = []) | Create a json response |
| success($message = "", $data = null) | Create a successful json response |
| error($message = "", $data = null) | Create a failed json response  |

### Testing

``` bash
composer test
```

### Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

### Security

If you discover any security related issues, please email info@witify.io instead of using the issue tracker.

## Credits

- [François Lévesque](https://github.com/francoislevesque)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `laravel-json-response` will be documented in this file

## 1.0.0 - 201X-XX-XX

- initial release


tests/JsonTraitTest.php
<?php

namespace Witify\LaravelJsonResponse\Tests;

use PHPUnit\Framework\TestCase;
use Witify\LaravelJsonResponse\JsonTrait;
use Witify\LaravelJsonResponse\ResponseFactory;

class JsonTraitTest extends TestCase
{
    public function setUp()
    {
        parent::setUp();
    }

    public function test_json_method_is_added()
    {
        $mock = $this->getMockForTrait(JsonTrait::class);
        $this->assertTrue(
            method_exists($mock, 'json'), 
            'Class does not have method json'
        );
    }

    public function test_json_method_returns_a_response_factory_object()
    {
        $mock = $this->getMockForTrait(JsonTrait::class);
        $this->assertTrue(
            method_exists($mock->json(), 'setStatusCode'), 
            'Class returned by json() does not have method setStatusCode'
        );
        $this->assertInstanceOf(ResponseFactory::class, $mock->json());
    }
}


tests/ResponseFactoryTest.php
<?php

namespace Witify\LaravelJsonResponse\Tests;

use PHPUnit\Framework\TestCase;
use Witify\LaravelJsonResponse\ResponseFactory;

class ResponseFactoryTest extends TestCase
{
    private $factory;

    public function setUp()
    {
        parent::setUp();
        $this->factory = new ResponseFactory();
    }

    public function test_default_status_code_is_200()
    {
        $this->assertEquals($this->factory->getStatusCode(), 200);
    }

    public function test_set_status_code()
    {
        $statusCode = 422;
        $this->factory->setStatusCode($statusCode);
        $this->assertEquals($this->factory->getStatusCode(), $statusCode);
    }

    public function test_json_response()
    {
        $data = ['testData' => 3];
        $response = $this->factory->response($data);
        $this->assertEquals($response->getStatusCode(), 200);
        $this->assertEquals($response->getData(), (object) $data);
    }

    public function test_json_response_success()
    {
        $response = $this->factory->success();
        $this->assertEquals($response->getStatusCode(), 200);
        $this->assertTrue($response->getData()->success);
        $this->assertObjectHasAttribute('message', $response->getData());
        $this->assertObjectHasAttribute('data', $response->getData());
    }

    public function test_json_response_error()
    {
        $response = $this->factory->error();
        $this->assertEquals($response->getStatusCode(), 200);
        $this->assertFalse($response->getData()->success);
        $this->assertObjectHasAttribute('errors', $response->getData());
        $this->assertObjectHasAttribute('data', $response->getData());
    }

    public function test_json_redirect()
    {
        $url = '/test';
        $response = $this->factory->redirect($url);
        $this->assertEquals($response->getStatusCode(), 200);
        $this->assertEquals($response->getData()->redirect, $url);
    }

    public function test_json_created()
    {
        $response = $this->factory->created();
        $this->assertEquals($response->getStatusCode(), 201);
    }

    public function test_json_bad_request()
    {
        $response = $this->factory->badRequest();
        $this->assertEquals($response->getStatusCode(), 400);
    }

    public function test_json_forbidden()
    {
        $response = $this->factory->forbidden();
        $this->assertEquals($response->getStatusCode(), 403);
    }

    public function test_json_not_found()
    {
        $response = $this->factory->notFound();
        $this->assertEquals($response->getStatusCode(), 404);
    }

    public function test_json_bad_entity()
    {
        $response = $this->factory->badEntity();
        $this->assertEquals($response->getStatusCode(), 422);
    }

    public function test_json_server_error()
    {
        $response = $this->factory->serverError();
        $this->assertEquals($response->getStatusCode(), 500);
    }
}



src/ResponseFactory.php
<?php

namespace Witify\LaravelJsonResponse;

use Illuminate\Http\JsonResponse;

class ResponseFactory
{
    /**
     * Http status code 
     *
     * @var string
     */
    private $statusCode = 200;

    /**
     * Get the http status code
     *
     * @return void
     */
    public function getStatusCode() : int
    {
        return $this->statusCode;
    }

    /**
     * Set the http status code
     *
     * @param int $statusCode
     * @return ResponseFactory
     */
    public function setStatusCode(int $statusCode) : ResponseFactory
    {
        $this->statusCode = $statusCode;
        return $this;
    }

    /**
     * Creates a json response
     *
     * @param mixed $data
     * @param array $headers
     * @return Response
     */
    public function response($data, $headers = [])
    {
        return new JsonResponse($data, $this->getStatusCode(), $headers);
    }

    /**
     * Success response
     *
     * @param string $message
     * @param mixed $data
     * @return Response
     */
    public function success($message = "", $data = null)
    {
        return $this->response([
            'success' => true,
            'message' => $message,
            'data' => $data,
        ]);
    }

    /**
     * Error response
     *
     * @param string $message
     * @param mixed $data
     * @return Response
     */
    public function error($message = "", $data = null)
    {
        return $this->response([
            'success' => false,
            'errors' => [
                'message' => $message,
                'status_code' => $this->getStatusCode(),
            ],
            'data' => $data,
        ]);
    }

    /**
     * Redirect response
     *
     * @param string $url
     * @param mixed $object
     * @return Response
     */
    public function redirect(string $url, $data = null)
    {
        return $this->response([
            'success' => true,
            'redirect' => $url,
            'data' => $data,
        ]);
    }

    /**
     * Created response
     *
     * @param string $message
     * @param mixed $data
     * @return Response
     */
    public function created($message = 'Successfully created', $data = null)
    {
        return $this->setStatusCode(201)->success($message, $data);
    }

    /**
     * Bad request response
     *
     * @param string $message
     * @param mixed $data
     * @return Response
     */
    public function badRequest($message = 'Bad Request', $data = null)
    {
        return $this->setStatusCode(400)->error($message, $data);
    }

    /**
     * Forbidden response
     *
     * @param string $message
     * @param mixed $data
     * @return Response
     */
    public function forbidden($message = 'Forbidden', $data = null)
    {
        return $this->setStatusCode(403)->error($message, $data);
    }

    /**
     * Not found response
     *
     * @param string $message
     * @param mixed $data
     * @return Response
     */
    public function notFound($message = 'Not Found', $data = null)
    {
        return $this->setStatusCode(404)->error($message, $data);
    }

    /**
     * Bad entity response
     *
     * @param string $message
     * @param mixed $data
     * @return Response
     */
    public function badEntity($message = 'Bad Entity', $data = null)
    {
        return $this->setStatusCode(422)->error($message, $data);
    }

    /**
     * Server error response
     *
     * @param string $message
     * @param mixed $data
     * @return Response
     */
    public function serverError($message = 'Internal Server Error', $data = null)
    {
        return $this->setStatusCode(500)->error($message, $data);
    }
}
